COLLABORATOR_GUIDE.md

```diff
@@ -84,6 +87,190 @@ All pull requests that modify executable code should be subjected to
 continuous integration tests on the
 [project CI server](https://ci.nodejs.org/).
 
+### Internal vs. Public API
+
+Due to the nature of the JavaScript language, it can often be difficult to
+establish a clear distinction between which parts of the Node.js implementation
+represent the "public" API Node.js users should assume to be stable and which
+are considered part of the "internal" implementation detail of Node.js itself.
+A general rule of thumb has been to based the determination off what
+functionality is actually *documented* in the official Node.js API
+documentation. However, it has been repeatedly demonstrated that either the
+documentation does not completely cover implemented behavior or that Node.js
+users have come to rely heavily on non-documented aspects of the Node.js
+implementation.
+
+While there a numerous exceptions, the following general rules should be
+followed to determine which aspects of the Node.js API are considered
+"internal":
+
+- Any and all functionality exposed via the `process.binding(...)` API is
+  considered to be internal and *not* part of the Node.js Public API.
+- Any and all functionality exposed in `lib/internal/**/*.js` that is not
+  re-exported by code in `lib/*.js` is considered to be internal.
```

Comment:
I think the distinction here lies on whether the functionality is documented as part of the external public API.